### PR TITLE
poly - Fixes the issue#19353 of simplified trace matrix error attribute

### LINKED ISSUE #19353: simplify(1/Trace(constant_matrix)) throws AttributeError
This is just like https://github.com/sympy/sympy/issues/19327, but unlike that issue happens with no local modifications:

```ipython
In [1]: from sympy import *

In [2]: m = Matrix([[30, 2], [3, 4]])

In [3]: simplify(1/Trace(m))
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-61-7221c8bb7585> in <module>
----> 1 simplify(1/Trace(m))

c:\users\wiese\repos\ga\sympy\sympy\simplify\simplify.py in simplify(expr, ratio, measure, rational, inverse, doit, **kwargs)
    600     expr = bottom_up(expr, lambda w: getattr(w, 'normal', lambda: w)())
    601     expr = Mul(*powsimp(expr).as_content_primitive())
--> 602     _e = cancel(expr)
    603     expr1 = shorter(_e, _mexpand(_e).cancel())  # issue 6829
    604     expr2 = shorter(together(expr, deep=True), together(expr1, deep=True))

c:\users\wiese\repos\ga\sympy\sympy\polys\polytools.py in cancel(f, *gens, **args)
   6664
   6665     try:
-> 6666         (F, G), opt = parallel_poly_from_expr((p, q), *gens, **args)
   6667     except PolificationFailed:
   6668         if not isinstance(f, (tuple, Tuple)):

c:\users\wiese\repos\ga\sympy\sympy\polys\polytools.py in parallel_poly_from_expr(exprs, *gens, **args)
   4310     """Construct polynomials from expressions. """
   4311     opt = options.build_options(gens, args)
-> 4312     return _parallel_poly_from_expr(exprs, opt)
   4313
   4314

c:\users\wiese\repos\ga\sympy\sympy\polys\polytools.py in _parallel_poly_from_expr(exprs, opt)
   4363             exprs[i] = exprs[i].as_expr()
   4364
-> 4365     reps, opt = _parallel_dict_from_expr(exprs, opt)
   4366     if not opt.gens:
   4367         raise PolificationFailed(opt, origs, exprs, True)

c:\users\wiese\repos\ga\sympy\sympy\polys\polyutils.py in _parallel_dict_from_expr(exprs, opt)
    332         reps, gens = _parallel_dict_from_expr_if_gens(exprs, opt)
    333     else:
--> 334         reps, gens = _parallel_dict_from_expr_no_gens(exprs, opt)
    335
    336     return reps, opt.clone({'gens': gens})

c:\users\wiese\repos\ga\sympy\sympy\polys\polyutils.py in _parallel_dict_from_expr_no_gens(exprs, opt)
    274         reprs.append(terms)
    275
--> 276     gens = _sort_gens(gens, opt=opt)
    277     k, indices = len(gens), {}
    278

c:\users\wiese\repos\ga\sympy\sympy\polys\polyutils.py in _sort_gens(gens, **args)
    101
    102     try:
--> 103         gens = sorted(gens, key=order_key)
    104     except TypeError:  # pragma: no cover
    105         pass

c:\users\wiese\repos\ga\sympy\sympy\polys\polyutils.py in order_key(gen)
     81                 pass
     82
---> 83         name, index = _re_gen.match(gen).groups()
     84
     85         if index:

AttributeError: 'NoneType' object has no attribute 'groups'
```

### PR DESCRIPTION
**Pull request by praneethratna**
#### Brief description of what is fixed or changed
Fixes #19353 by changing the sympy/polys/polyutils.py file. In this file the line changed is
from this  _re_gen = re.compile(r"^(.+?)(\d*)$")
to this  _re_gen = re.compile(r"^(.*?)(\d*)$", re.MULTILINE)
A test case function test_issue_19353() has also been added in sympy/polys/tests/test_polyutils.py
#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
A test for the error with trace would be good as well.

**Comment by praneethratna** [comment]
> A test for the error with trace would be good as well.

yea , on it.

**Comment by praneethratna** [comment]
@oscarbenjamin i've updated the tests , is it fine now?

File: sympy/polys/tests/test_polyutils.py [hunk 1]
```diff
@@ -286,3 +286,9 @@ def test_dict_from_expr():
         ({(0,): -Integer(1), (1,): Integer(1)}, (x,))
     raises(PolynomialError, lambda: dict_from_expr(A*B - B*A))
     raises(PolynomialError, lambda: dict_from_expr(S.true))
+
+
+def test_issue_19353():
+    m = Matrix([[30, 2], [3, 4]])
+    assert Poly(Symbol('\n1'))
+    assert simplify(1/Trace(m)) == Rational('1/34')
```
Comment (oscarbenjamin) [inline sympy/polys/tests/test_polyutils.py:294]: We should avoid unnecessary string parsing in the tests so it should be `Rational(1, 34)`.

Also `assert Poly(Symbol('\n1'))` looks strange so it at least needs a comment explaining that it used to raise an exception and there shouldn't be an assert in that case.

I think it's better to separate these two tests and put them along with other similar tests. I think that the poly error is to do with `_sort_gens` so the `Poly` test could be added to `test__sort_gens` above and written like the others. Something like:
```
# https://github.com/sympy/sympy/issues/19353
n1 = Symbol('\n1')
assert _sort_gens([x, n1]) == ...
```
The test involving trace could be added to either the matrix expression tests or the simplify tests if there is already a place with similar tests.

Comment (praneethratna) [reply sympy/polys/tests/test_polyutils.py]: @oscarbenjamin made the changes accordingly , also I've added the test involving trace in the test_simplify() function in matrices tests.

File: sympy/matrices/tests/test_commonmatrix.py [hunk 2]
```diff
@@ -561,6 +561,9 @@ def test_simplify():
     assert M.simplify() == Matrix([[eq]])
     assert M.simplify(ratio=oo) == Matrix([[eq.simplify(ratio=oo)]])
 
+    # https://github.com/sympy/sympy/issues/19353
+    m = Matrix([[30, 2], [3, 4]])
+    assert (1/(m.trace())).simplify() == Rational(1, 34)
 
 def test_subs():
```
Comment (oscarbenjamin) [inline sympy/matrices/tests/test_commonmatrix.py:568]: There should be two blank lines between top-level functions.

**Comment by smichr** [comment]
Thanks, this is in.

**Pull request merged**